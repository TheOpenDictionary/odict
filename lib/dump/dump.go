package dump

import (
	"github.com/TheOpenDictionary/odict/lib/types"
	"github.com/TheOpenDictionary/odict/lib/utils"
)

type SqlDialect = string

const (
	Postgres  SqlDialect = "postgres"
	Sqlite    SqlDialect = "sqlite"
	Mysql     SqlDialect = "mysql"
	Sqlserver SqlDialect = "sqlserver"
)

// AsXML converts an Dictionary struct
// to its original ODXML
func AsXML(dict *types.Dictionary) (string, error) {
	return utils.SerializeToXML(dict.AsRepresentable(), true)
}

// AsSQL converts an Dictionary struct
// to SQL output to seed a database
func AsSQL(dict *types.Dictionary, sqlDialect SqlDialect) (string, error) {
	return sql(dict.AsRepresentable(), sqlDialect)
}
