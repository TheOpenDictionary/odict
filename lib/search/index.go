package search

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"time"

	"github.com/TheOpenDictionary/lib/config"
	"github.com/TheOpenDictionary/odict/lib/types"
	"github.com/blevesearch/bleve/v2"
	"github.com/blevesearch/bleve/v2/document"
	"github.com/blevesearch/bleve/v2/index/scorch"
	idx "github.com/blevesearch/bleve_index_api"
	"github.com/schollz/progressbar/v3"
)

type IndexRequest struct {
	Dictionary *types.Dictionary
	Overwrite  bool
	Quiet      bool
}

func getIndexPath(dictionaryID string) (string, error) {
	path, err := config.GetConfigDir()

	if err != nil {
		return "", err
	}

	return filepath.Join(path, "idx", dictionaryID), nil
}

func Index(request IndexRequest) (string, error) {
	dictionary := request.Dictionary.AsRepresentable()
	indexPath, err := getIndexPath(dictionary.ID)

	if err != nil {
		return "", err
	}

	_, statErr := os.Stat(indexPath)

	if os.IsNotExist(statErr) {
		if !request.Quiet {
			fmt.Println("Indexing dictionary (this might take some time)...")
		}

		mapping := bleve.NewIndexMapping()

		index, indexErr := bleve.NewUsing(indexPath, mapping, scorch.Name, scorch.Name, nil)

		if indexErr != nil {
			return "", indexErr
		}

		defer index.Close()

		totalEntries := len(dictionary.Entries)

		var bar *progressbar.ProgressBar

		if !request.Quiet {
			bar = progressbar.Default(int64(totalEntries))
		}

		batch := index.NewBatch()
		batchCount := 0
		batchSize := 100

		for key := range dictionary.Entries {
			entry := dictionary.Entries[key]
			doc := document.NewDocument(strings.ToLower(entry.Term))

			mapping.MapDocument(doc, entry)

			enc := types.Serialize(&entry)
			field := document.NewTextFieldWithIndexingOptions("_source", nil, enc, idx.StoreField)
			nd := doc.AddField(field)

			batch.IndexAdvanced(nd)

			batchCount++

			if !request.Quiet {
				bar.Add(1)
			}

			time.Sleep(time.Millisecond)

			if batchCount >= batchSize {
				idxErr := index.Batch(batch)

				if idxErr != nil {
					return "", idxErr
				}

				batch = index.NewBatch()
				batchCount = 0
			}
		}

		idxErr := index.Batch(batch)

		if idxErr != nil {
			return "", idxErr
		}

		if !request.Quiet {
			fmt.Println()
		}
	} else if request.Overwrite {
		if !request.Quiet {
			fmt.Println("Purging existing index...")
		}

		os.RemoveAll(indexPath)

		return Index(IndexRequest{Dictionary: request.Dictionary, Overwrite: false, Quiet: request.Quiet})
	}

	return indexPath, nil
}
