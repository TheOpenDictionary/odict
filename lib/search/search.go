package search

import (
	"fmt"
	"log"
	"os"
	"strings"

	"github.com/TheOpenDictionary/odict/lib/types"
	"github.com/TheOpenDictionary/odict/lib/utils"
	"github.com/blevesearch/bleve/v2"
	query "github.com/blevesearch/bleve/v2/search/query"
)

// SearchDictionary searches a dictionary model using Bleve using
// it's unique dictionary ID
func SearchDictionary(dictionaryID string, queryStr string, exact bool) []types.Entry {
	indexPath := getIndexPath(dictionaryID)
	_, err := os.Stat(indexPath)

	if os.IsNotExist(err) {
		log.Fatalln("Index path does not exist. Did you call IndexDictionary() first?")
	}

	index, openErr := bleve.Open(indexPath)

	utils.Check(openErr)

	defer index.Close()

	var query query.Query = bleve.NewMatchQuery(queryStr)

	if exact {
		query = bleve.NewDocIDQuery([]string{strings.ToLower(queryStr)})
	}

	search := bleve.NewSearchRequest(query)
	search.Fields = []string{"_source"}
	searchResults, searchErr := index.Search(search)

	utils.Check(searchErr)

	hits := searchResults.Hits

	entries := make([]types.Entry, len(hits))

	for i, x := range hits {
		b, ok := x.Fields["_source"]

		if ok {
			entries[i] = *types.GetRootAsEntry([]byte(fmt.Sprintf("%v", b)), 0)
		}
	}

	return entries
}
