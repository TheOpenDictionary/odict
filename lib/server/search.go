package server

import (
	"encoding/json"
	"net/http"

	"github.com/TheOpenDictionary/odict/lib/search"
	"github.com/TheOpenDictionary/odict/lib/types"
)

func handleSearch(pathOrAlias string) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		dictionary, err := getDictionary(pathOrAlias, r)

		if err != nil {
			http.Error(w, err.Error(), http.StatusBadRequest)
			return
		}

		query := r.URL.Query()

		queries := query.Get("query")

		entries, err := search.SearchDictionary(search.SearchDictionaryRequest{
			Dictionary: dictionary,
			Query:      queries,
			Exact:      false,
		})

		if err != nil {
			http.Error(w, err.Error(), http.StatusBadRequest)
			return
		}

		representable := types.EntriesToRepresentables(entries)

		// Return the result as JSON
		w.Header().Set("Content-Type", "application/json")

		json.NewEncoder(w).Encode(representable)
	}
}
