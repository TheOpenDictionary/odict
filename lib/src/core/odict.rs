/* -------------------------------------------------------------------------- */
/*                               DictionaryFile                               */
/* -------------------------------------------------------------------------- */

use std::path::PathBuf;

use rkyv::access_unchecked;

use crate::{semver::SemanticVersion, ArchivedDictionary, Dictionary};

#[derive(Clone)]
pub struct OpenDictionary {
    pub signature: String,
    pub version: SemanticVersion,
    pub path: Option<PathBuf>,
    pub total_size: u64,

    pub(super) content: Vec<u8>,
}

impl OpenDictionary {
    pub fn access(&self) -> crate::Result<&ArchivedDictionary> {
        let archived = unsafe { access_unchecked::<crate::ArchivedDictionary>(&self.content[..]) };
        Ok(archived)
    }

    pub fn deserialize(&self) -> crate::Result<Dictionary> {
        let dict: Dictionary = self.access()?.deserialize()?;
        Ok(dict)
    }
}
