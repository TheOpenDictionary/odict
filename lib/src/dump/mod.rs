mod json;
mod xml;

pub use self::json::*;
pub use self::xml::*;
