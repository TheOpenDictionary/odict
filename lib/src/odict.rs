/* -------------------------------------------------------------------------- */
/*                               DictionaryFile                               */
/* -------------------------------------------------------------------------- */

use std::path::PathBuf;

use crate::version::SemanticVersion;

#[derive(Clone)]
pub struct OpenDictionary {
    pub(crate) signature: String,
    pub(crate) version: SemanticVersion,
    pub(crate) path: Option<PathBuf>,
    pub(crate) bytes: Vec<u8>,
}

impl OpenDictionary {
    pub fn signature(&self) -> &str {
        &self.signature
    }

    pub fn path(&self) -> &Option<PathBuf> {
        &self.path
    }

    pub fn version(&self) -> &SemanticVersion {
        &self.version
    }

    pub fn size(&self) -> crate::Result<u64> {
        if let Some(path) = &self.path {
            if let Ok(metadata) = std::fs::metadata(path) {
                return Ok(metadata.len());
            }
        }

        let bytes = self.to_bytes()?;

        Ok(bytes.len().try_into()?)
    }

    pub fn contents(&self) -> crate::Result<&crate::schema::ArchivedDictionary> {
        Ok(rkyv::access::<
            crate::schema::ArchivedDictionary,
            rkyv::rancor::Error,
        >(&self.bytes[..])?)
    }
}
