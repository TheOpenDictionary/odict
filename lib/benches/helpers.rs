use std::sync::LazyLock;

use odict::{DictionaryFile, DictionaryReader, DictionaryWriter, Result};

pub fn create_archive_dict(name: &str) -> Result<DictionaryFile> {
    let reader = DictionaryReader::default();
    let writer = DictionaryWriter::default();
    let input = format!("../examples/{}.xml", name);
    let output = format!("../examples/{}.odict", name);

    writer.compile_xml(&input, &output)?;
    reader.read_from_path(&output)
}

pub static EXAMPLE_DICTIONARY_1: LazyLock<DictionaryFile> =
    LazyLock::new(|| create_archive_dict("wiktionary").unwrap());
