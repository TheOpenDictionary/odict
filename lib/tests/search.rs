mod helpers;

#[cfg(test)]
mod index_tests {

    use super::helpers::EXAMPLE_DICTIONARY_1;
    use insta::assert_snapshot;
    use odict::dump::ToJSON;
    use odict::search::{IndexOptions, SearchOptions};

    #[test]
    fn test_search() {
        let dict = EXAMPLE_DICTIONARY_1.to_dictionary().unwrap();
        let dir = ".odict/.idx";

        dict.index(IndexOptions::default().dir(dir)).unwrap();

        let result = dict.search("a dog", SearchOptions::default().dir(dir));

        assert_eq!(result.is_err(), false);
        assert_snapshot!(result.unwrap().to_json(true).unwrap());
    }
}
