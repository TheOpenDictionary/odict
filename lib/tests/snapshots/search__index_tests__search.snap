---
source: lib/tests/search.rs
expression: result.unwrap().to_json(true).unwrap()
---
[
  {
    "term": "dog",
    "etymologies": [
      {
        "pronunciation": "dooooog",
        "description": "Latin root",
        "senses": {
          "un": {
            "pos": "un",
            "definitions": [
              {
                "type": "definition",
                "value": "a dog"
              }
            ]
          }
        }
      }
    ]
  }
]
